### DISCUSSION TITLE
Podcast Implementation

### DISCUSSION THREAD

**Discussion by jflattery** [advplyr/audiobookshelf#408] [Ideas]
What would a podcast Implementation look like to you? What features would you like to see?

I'll start:

- AudioBookShelf (ABS) admins, should be able to allow/prevent end users from adding their own podcasts.

- ~_Automatically downloads latest podcast episodes._~ (**Implemented**)
    - Custom fetch schedule per podcast (possibly fuzzy, IE if a podcast has a release schedule of "at a random TOD Bi-Weekly on Tuesdays")
 
- ~_The application should support custom remote subscription URLs_~ (**Implemented**)
    - ~_The application should be able to access feed protected by username and password [Original Request](https://github.com/advplyr/audiobookshelf/discussions/408#discussioncomment-2462670)_~ (**Implemented**)

- The application should hide remote subscription URLs from end user, unless they are the one who added the subscription.

- ~_Might be nice if like [podgrab](https://github.com/akhilrex/podgrab), the application hosted podcast to be downloaded from the user's preferred client (overcast, downcast, etc)_~ (**Implemented**)

- Would be AWESOME if it supported [SponsorBlock](https://sponsor.ajay.app/). [Comment by the dev of SponsorBlock on Podcast Implementation](https://old.reddit.com/r/SponsorBlock/comments/g4e316/sponsorblock_for_podcasts/).

- Support ID3 Chapters

**Comment by timbatog** [comment]
I second this. I’d very definitely financially contribute to this being implemented. 


10 Mar 2022, 14:02 by ***@***.***:

**Comment by ItsNoted** [comment]
I opened an [enhancement request](https://github.com/advplyr/audiobookshelf/issues/428) already but I'll add it here too just in case. Showing the date when the episode was added on podcast pages like seen in the screenshot below.

![image](https://user-images.githubusercontent.com/57927413/160627479-b8b0b0d7-cd88-42e9-8857-6f5149cd505b.png)

**Reply by Zibbp** [reply]
Another possible UI addition could be if you click the episode it drops down an area below the episode card where the episode description/notes can be read. Currently it can only be viewed when editing the episode.

**Comment by freitagdavid** [comment]
Support for authenticated podcasts would be great. The best example I can think of for this is Mysterious Universe.

**Reply by jflattery** [reply]
To confirm, this a feed which requires a username & password?

**Reply by freitagdavid** [reply]
Yep, so it would be a podcast with a paid subscription. So for example Mysterious Universe has a free feed where you get an episode a week, you can pay and you get 2 episodes a week and the normal show is extended.

**Reply by advplyr** [reply]
I did a quick look into this and the username & password can be put directly into the RSS feed for certain podcasts.

`http://user:pass@somesite.com/feed`

Is that not the case for Mysterious Universe?

**Reply by jflattery** [reply]
@freitagdavid Can you confirm if the format `http://user:pass@somesite.com/feed` will work for Mysterious Universe?

**Comment by Zibbp** [comment]
Adding support during the library scan to read ID3 tags would be fantastic. Currently if episodes are imported the ID3 tags do not appeared to be read and used. Below is an image of downloading some episodes via the UI then copying those episodes and doing a library scan to import them. None of the tags appear to be read and used for populating.

(Downloaded via UI)
![image](https://user-images.githubusercontent.com/21207065/160660919-4fe6b671-37b3-4820-9a6f-00f879aab66d.png)

(Copy the above downloaded ones to a new folder and import via library scan)
![image](https://user-images.githubusercontent.com/21207065/160661017-90255e45-9285-437b-8bf8-e8e9c59dab1a.png)

**Reply by advplyr** [reply]
This should be easy since we are already parsing all that info and it is stored with the Audio File.
Do you have an idea for a mapping like for audiobooks: https://www.audiobookshelf.org/docs#metadata ?

**Reply by Zibbp** [reply]
These are the tags for an Episode of Darknet Diaries I downloaded via yt-dlp directly form the rss feed.
![image](https://user-images.githubusercontent.com/21207065/160872790-556ab24c-b7f2-4eb7-bfde-60a800a20174.png)

Comparing that file versus the same episode downloaded from the UI, there are a few differences in the fields.
![image](https://user-images.githubusercontent.com/21207065/160872953-4feee472-1849-415b-849b-d77e622dc066.png)

Does anyone else have downloaded podcast files that can post what the tags are for comparison?

**Reply by Catsrules** [reply]
@Zibbp I am trying to play around with importing podcasts and viewing the ID3 data, however your screenshots look different then mine. How did you get to the editing part of Episode and Episode Type and Pub Date? Maybe I am just missing something but mine looks the same as it does with Audiobook. And All podcast files are combined into one "book".

I have even created a new library with the podcast type but it still seems to be treated the same as book.

**Reply by jflattery** [reply]
> @Zibbp I am trying to play around with importing podcasts and viewing the ID3 data, however your screenshots look different then mine. How did you get to the editing part of Episode and Episode Type and Pub Date? Maybe I am just missing something but mine looks the same as it does with Audiobook. And All podcast files are combined into one "book".
> 
> 
> 
> I have even created a new library with the podcast type but it still seems to be treated the same as book. 

The screenshots are from a pre-release build of v2 which does not currently work with the mobile apps that are currently available.

**Reply by Catsrules** [reply]
Ahh that would explain why can't find it :).

Thanks.

**Comment by MidnightSnowleopard** [comment]
@Zibbp  Audiobooshelf already has the ability to use IDv3 tags from files.  You do need to go into the settings and enable the setting named "Scanner prefer audio metadata" for it to read them.  It is not enabled by default

**Reply by Zibbp** [reply]
I've tried enabling that option but for me it does not work with podcasts on the master branch.

**Comment by advplyr** [comment]
Podcast support is official in [v2.0.1](https://github.com/advplyr/audiobookshelf/releases/tag/v2.0.1) although just the beginning!

**Comment by msfoote** [comment]
Something that would be awesome for the podcast support would be something akin the Queue feature in the AntennaPod app that I use on Android.

![image](https://user-images.githubusercontent.com/1077698/164950789-074d33f2-1fb2-4bb8-8d85-e28218795692.png)

This is a place where I can listen to all of my not-listened podcast episodes in the order that they dropped. As you can see above it starts with an episode that aired on Apr 12 and then one on the 14th.  In this particular app I can reorder them if I wish to and so I have the 19th episode queued prior to the April 17th.

**Comment by advplyr** [comment]
Podcast update incoming in v2.0.2 that will show currently downloading episodes and number of episodes in download queue.

![image](https://user-images.githubusercontent.com/67830747/164951098-e5e2e69c-aaee-460f-84c9-03022307a953.png)

**Reply by Jdiesel87** [reply]
Just giving this a test. It wasn't something I thought I needed but after setting it up I could see myself using it.

Question: Are there any downloaders/grabbers that support Spotify? I have a couple Spotify exclusive podcasts that I listen to and not having everything in one app is a bit of a deterrent for me right now.

**Comment by backcountrymountains** [comment]
Cool Project!

I'm also trying to use this for news podcasts. My ideal view is a list of all podcasts sorted by most recent episodes. Additionally, It would be nice to see the detailed description of each episode in the list view. Also, the ability to queue podcast episodes in the player would be nice. 

I don't know how to modify the code to be able to make my own view. I have a vague understanding of HTML but I don't see how the page is generated. Is there a tutorial, besides a 6-month web development bootcamp, that anyone can recommend for working with a project like this?

Thanks.

**Reply by msfoote** [reply]
I echo this as well.

**Reply by advplyr** [reply]
We plan on making a queue for the player. As for the podcast episode view there is a lot of improvements we need to make here but what we really need is someone helping with the UI/UX design. Even if it is just some mockups in paint or you can point to some other apps/software that have the design you are looking for.

**Reply by backcountrymountains** [reply]
I'm excited!
I've attached a super-refined (not) mock-up I did in MS Paint. I'd just like a list, sorted by pubDate, showing all of my news podcast episodes with their titles and descriptions so I can quickly see which ones interest me. And maybe a little star button to highlight the episodes that I'd like to listen to later.

I'm excited! Thanks!
![Audiobookshelf Podcast view](https://user-images.githubusercontent.com/29560911/187263157-1dfd45cf-4f3a-4e80-871c-7b00ac44f6fd.png)

**Reply by Catsrules** [reply]
One thing I would mention with the description is a maximum character length. As some of these podcast have huge descriptions, with sponsor, show notes, credits etc.. I think it would be to much information for a summery page to have. If you want to full description you would need to interact with the episode in some way. Maybe with a dedicated button or just click/tap on the episode 


Since we are on the topic of suggestions. One thing I have been looking for is a "Next Episode" or "Next Up" section on the Home Screen. 

For example if I finish episode 4 of a podcast episode 5 should now be listed on the Home Screen under the Next Episode section.  As some podcasts you want to listen to them in order and you might get behind so you don't want to listen to the Newest Episode yet.

**Reply by advplyr** [reply]
The full description is going to be an issue because of what @Catsrules said. Podcast RSS feeds typically come with a summary and a full description, but not all podcasts do this. The summary is meant to be plaintext and limited in length whereas the full description can be very long.

@Catsrules  The "Next Episode" category on the home page is requested here #861 and I will probably get to this in the next release.

I was able to implement the auto-play next episode and a queue for the audio player for the web apps next release. Will put this in the mobile app next.

Also, I'm not sure about the star system will need to think about that some more.

**Reply by backcountrymountains** [reply]
I would definitely be happy with podcast summary instead of description. There is a lot of extraneous stuff in description. I can also see how, if the summary isn't present, being able to click/tap to show the description would be nice.

A list view that sorts all podcasts by pubDate (I can't find this as an option in library view) and shows the summary for each podcast would be 90% of what I could ever want. To me, it seems that having the individual podcast view work as a whole library view is pretty close to what I'm looking for.

Extra stuff would be deleting old episodes, saving starred episodes and extra player functionality for queuing. 

I'm also realizing that these enhancements are focused on daily news-type podcasts and I might need to make a separate library for weekly-type podcasts with some kind of "group by what's new this week" sort.

Anyway, this project is great!

**Reply by advplyr** [reply]
The podcast episode list already functions like that with the summary and click on it to view the full description. I don't think the full description is implemented on mobile apps yet, just the summary there.

You can also already delete old episodes but you have to do it 1 by 1. If you have auto-download setup you can specify the max number of episodes you want to keep which will automatically remove an episode if you go beyond that max number.

**Reply by backcountrymountains** [reply]
Yes, the podcast episode list view is exactly what I want my podcast library view to be! So, instead of having to pick a podcast and look at all the episodes, I'd like to be able to see all of the most recent episodes in my whole podcast library at the same time, with the same view as the individual podcast episode list. Then I can select and queue up several different podcasts to listen to the news while I walk the dog. I'm not saying it's an exciting use case.

It's pretty annoying that each podcast has their own idea about what goes in summary, if they even use it, and what extraneous crap goes in the description.

I didn't see that max episode setting before! Thanks for pointing it out!

**Reply by advplyr** [reply]
Ah I understand now, that makes sense.
